Replace router support for ingress with an ingress-to-route controller

smarterclayton opened the pull request:
The existing ingress support in the router has a number of serious problems:

1. We cannot give a router "watch all secrets safely", so secret updates are delayed to the full resync interval
2. Granting a router access to individual secrets requires some sort of ingress aware controller
3. There are race conditions between secret update where a secret may not change for a full resync interval
4. Users get no status on ingress, and the UI cannot show anything about them, which means cert and overlap failures are opaque

This PR changes from the routers handling ingress to a new `ingress-to-route` controller that creates actual routes for each ingress, then keeps them in sync. It also has permission to watch secrets and will update the routes when those change, which means routers observe ingress secret changes as quickly as possible.  Each child route has an owner/controller ref back to the ingress, which means garbage collection manages its lifecycle.

Example:

```
kind: Ingress
apiVersion: extensions/v1beta1
metadata:
  name: test
spec:
  rules:
  - host: test.com
     http:
       paths:
       - path: /test 
         backend:
           serviceName: test-1
           servicePort: 80

-->

kind: Route
apiVersion: route.openshift.io/v1
metadata:
  name: test-a34th
  ownerReferences:
  - apiVersion: extensions/v1beta1
     kind: Ingress
     name: test
     controller: true
spec:
  host: test.com
  path: /test
  to:
    name: test-1
  port:
     targetPort: 80
```

It is also much cleaner, easier to test (unit and in isolation), easier to review, and simplifies router code significantly. I had originally looked at implementing this as an RBAC sync controller (granting routers access to secrets) but that only exposed the real problem which was that routers can't watch on secrets. Since Routes are generally better designed as a secure API than ingress, and ingress is likely to substantially change over the next few releases, this approach allows us to continue to shield our users from the impact of that uncertainty while preserving all the good things about routes.

Supporting ingress correctly is a key short term goal so that we can adapt users in the ecosystem who depend on ingress.

This adds no feature support above for ingress above and beyond, but it is relatively easy for us to adapt community annotations that can use our features.

Still remaining:

* [x] Add an e2e test that verifies the router serves a materialized ingress
* [x] Determine how/whether to default spec.host (arguably we shouldn't be supporting this since ingress controllers don't have this concept)
* [x] Decide whether to expose load balancer status

smarterclayton commented:
@openshift/sig-networking @liggitt re: our security discussion.  Implementing a full secret cache a la the Kubelet is not worth it.  Route API is better designed for actual security than ingress in this case.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
@ramr

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+	eventRecorder record.EventRecorder
+
+	// Allows injection for testing, controls requeues on errors
+	resourceFailureDelayFn func(requeue int) (time.Duration, bool)
```
dead?

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+
+// defaultResourceFailureDelay will retry failures forever, but implements an exponential
+// capped backoff after a certain limit.
+func defaultResourceFailureDelay(requeue int) (time.Duration, bool) {
```
which would make this dead

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+		ingressLister: ingresses.Lister(),
+		secretLister:  secrets.Lister(),
+		routeLister:   routes.Lister(),
+		serviceLister: services.Lister(),
```
normally, you have a sync functions collected from these for a "wait for sync before run" check.

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+		resourceFailureDelayFn: defaultResourceFailureDelay,
+	}
+
+	processNamespaceFn := func(obj interface{}) {
```
These are usually separate queuing functions off the Controller struct.

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+			case *v1.Secret:
+				return t.Type == v1.SecretTypeTLS
+			}
+			return true
```
if you're not a secret, why bother queuing.  It won't work right.

smarterclayton replied:
tombstones?

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+	}
+
+	processNamespaceFn := func(obj interface{}) {
+		switch t := obj.(type) {
```
`meta.Accessor` is general case, right? It feels really close to a general pattern.  I tried once before.  I don't remember what stopped it.

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    processNamespaceFn,
+			DeleteFunc: processNamespaceFn,
```
your function handles tombstones?  I didn't realize they were metav1.objects

smarterclayton replied:
we queue a tombstone by assuming that anything not a secret is probably a tombstone.  It didn't feel like handling it specially added anything over "if we don't recognize it, queue anyway"

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+		},
+	})
+
+	c.syncs = []cache.InformerSynced{
```
Ah, I see.  Yeah, reasonable.

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
```
We don't usually do this.  The last AddRateLimited is order of tens of minutes.  I keeps us from having to worry about permafails

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+
+func (c *Controller) sync(key queueKey) error {
+	// sync all ingresses in the namespace
+	if len(key.name) == 0 {
```
Why not do this when you enqueue?  Self feeding like this is unusual.

smarterclayton replied:
I guess no reason.

smarterclayton replied:
I remember, because it collapses all namespace retries for these entries.  Instead of each change traversing the cache, rapid changes result in de-duping all changes to the namespace into one, which then gets queued.  Both secrets and services can go through burst creation.  So instead of queuing everything in each namespace multiple times, we queue the namespace key.  I think this is a better pattern when the controller has to react to multiple blanket events.

smarterclayton replied:
I could clean the terminology up, but it makes the tests a lot more centralized

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+
+	ingress, err := c.ingressLister.Ingresses(key.namespace).Get(key.name)
+	if err != nil {
+		if errors.IsNotFound(err) {
```
Do this above the err check to reduce nesting

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+	processRouteFn := func(obj interface{}) {
+		switch t := obj.(type) {
+		case *routev1.Route:
+			name, ok := hasIngressOwnerRef(t.OwnerReferences)
```
`ingressName` please.  My eyes skipped past this detail

deads2k commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,591 @@
+	}
+
+	if len(errs) > 1 {
+		return fmt.Errorf("multiple errors occurred while creating the ingress: %v", errs)
```
`NewAggregate`?

deads2k commented:
@smarterclayton I didn't go down into the details, but this does read like a pretty simple controller.  I would queue directly instead of sometimes queuing an invalid key that you expand in your sync func though.

smarterclayton commented:
/test gcp

smarterclayton commented:
There appears to be a race condition in the router storage where a rapidly updated route is being seen as conflicting with itself (it could be the controller executing multiple writes but I don't see any evidence of that happening).  

That led me to identify an inconsistency in how we handle rejections inside the router - we should be passing down to our nested admission chain delete events when we reject an input, even if we may not have admitted it before (can leave old routes in the router state).  Will continue to test, but highlights that we need to have a consistency test for the router before we exit 1.10.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
I think the correct thing to do about ingress default routes is to enable that on routes first, then on ingress.  Since we already have rules around `spec.host` (it will be autofilled by the server), we'll need to add a new field to route v1 for `useDefaultHost bool` that skips spec.host filling in, then lets the router handle it.  Ingress would then set that in the absence of a spec.host value.  If a user wants auto generated route names... they can darn well use routes.  We'll probably need a security rule for what default host means - if i have `mycompany.com`, should all users who can add `foo.mycompany.com` also be able to add `mycompany.com/bar`?  It's a useful feature (as shown by ingress), but the security case needs to be settled.

Load balancer status can be added later if necessary.

smarterclayton commented:
If you agree with my reasoning about default ingress, this is ready for review as a real thing.

smarterclayton commented:
/retest

smarterclayton commented:
Test is flaking, looks like it conflicts with itself.  Either we’re creating two routes or this is a router state bug.  The fact that it occurs on _install indicates it isn’t just due to an older image.

smarterclayton commented:
/test install

smarterclayton commented:
/test install

smarterclayton commented:
/test install

smarterclayton commented:
/test extended_conformance_install

smarterclayton commented:
/retest

smarterclayton commented:
We don't support the router value defaulting on ingress correctly today, so I wanted to punt that to a follow up where we add a boolean on route spec like `useDefaultHost` or something which lets the router decide, and if the router isn't enforcing a default hostname it would be served on the wildcard domain.  That would make the router then compatible with ingress, although I think we need to figure out how to access control that.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

Miciah commented:
FWIW, I tested this in a devenv (`oc cluster up` and then `oc create -f /data/src/github.com/openshift/origin/test/extended/testdata/ingress.yaml`) and am seeing that raciness you mention.  For example, the ingress-to-route controller created two routes, "test-52lll" and "test-9kgxp", for host 2.ingress-test.com; "test-52lll" got admitted, "test-9kgxp" got denied ("HostAlreadyClaimed"), then "test-52lll" got deleted, and 10 minutes later (on the next router sync), "test-9kgxp" finally got admitted.

smarterclayton commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,548 @@
+	case metav1.Object:
+		ns := t.GetNamespace()
+		if len(ns) == 0 {
+			utilruntime.HandleError(fmt.Errorf("object %T has no namespace", obj))
```
All we care about here is the type of object that we don't recognize.

smarterclayton commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,548 @@
+	r.Spec.To.Weight = existing.Spec.To.Weight
+	if r.Spec.TLS != nil && existing.Spec.TLS != nil {
+		r.Spec.TLS.CACertificate = existing.Spec.TLS.CACertificate
+		r.Spec.TLS.DestinationCACertificate = existing.Spec.TLS.DestinationCACertificate
```
You can't set the destination in an ingress, so this was really just a sop left to allow someone to still get that feature.  We could add the more complex logic to reset based on type.

smarterclayton commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,548 @@
+		return tlsConfig == nil
+	}
+	if secret.Type != v1.SecretTypeTLS {
+		return tlsConfig == nil
```
This is to avoid allocating in the common case.  This controller gets woken up on any secret change in a namespace which is expected to be common, therefore we have to be very careful not to over allocate during evaluation.

Miciah replied:
I don't follow.  Is it common that
* an ingress exists,
* a corresponding route exists,
* the ingress references a secret,
* and that secret exists,
* but the secret is the wrong type?

If this is a case worth optimizing, would it also make sense to return true in the case that the secret does not exist (i.e., `return true` from `routeMatchesIngress` if `secretLister.Secrets(ingress.Namespace).Get(name)` is `nil`)? Why treat secret-does-not-exist differently from secret-exists-but-has-the-wrong-type (which probably means the secret does not exist, and another secret is using its name)?

smarterclayton replied:
Any change to a namespace for a secret will cause us to need to reevaluate every ingress in the namespace.  This is a level driven controller, so we will run over this code path for every ingress

1. everytime a secret changes in a namespace
2. everytime a service changes in a namespace
3. everytime a resync interval hits

So it does matter to be as efficient as possible because this line of code will be invoked quite often.

The reason we return sales is because the method does what it says on the tin - you can't reference a non TLS secret type from an ingress, so returning false means "doesn't match".

smarterclayton commented on pkg/route/controller/ingress/ingress.go:
```diff
@@ -0,0 +1,548 @@
+// controller creates a controller owner reference from the route to the parent ingress,
+// allowing users to orphan their ingress. All owned routes have specific spec fields
+// managed (those attributes present on the ingress), while any other fields may be
+// modified by the user.
```
"Managed" means "overwritten by the controller whenever the ingress changes".

smarterclayton commented on test/extended/util/framework.go:
```diff
@@ -273,6 +273,23 @@ func DumpPodLogs(pods []kapiv1.Pod, oc *CLI) {
+	}
+	for name, stdout := range values {
+		stdout = strings.TrimSuffix(stdout, "\n")
+		e2e.Logf(name + ": " + strings.Join(strings.Split(stdout, "\n"), fmt.Sprintf("\n%s: ", name)))
```
Not sure how it's safer.

Miciah replied:
In the off chance there's (e.g.) a `%s` in the output.  Granted the worst case is probably that you get some spurious "%!s(MISSING)" in the log output, but why wouldn't you just say what you mean?

smarterclayton commented:
Updated with comments.  I reproduced the issue on the server side that caused multiple route creations and deletions to be recorded and fixed that in the last commit (like other controllers that generate resources that don't have fixed names, we needed to use an expectation to avoid racing with the caches).

There is indeed a real problem in the router that it looks like a stress test should be able to reproduce.  I suspect it is due to improperly clearing an underlying cache entry for duplicate names.  Going to spawn a separate issue for that and ensure that whatever stress infrastructure Ram and I create that it correctly finds that issue.

smarterclayton commented:
/test extended_conformance_k8s

smarterclayton commented:
After looking at unique_host, it looks like route claiming is broken when paths are involved.  I'm going to open a separate PR that fixes the issues.  At heart, when a route is deleted other valid routes aren't being restored.  This looks like a bug that we've had for a very long time.

smarterclayton commented:
Fixed the unique_host issue in #19175.

knobunc commented:
/lgtm
